### PR TITLE
FIX: Fix for isolation level for in-built connection pool

### PR DESCRIPTION
**Pull request by subrata-ms**
> [AB#40573](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/40573)

> GitHub Issue: #337 

-------------------------------------------------------------------
### Summary   
  
This pull request addresses a critical issue in connection pooling for the `mssql_python` library by ensuring that the transaction isolation level is properly reset when a pooled connection is reused. Previously, the isolation level could leak between usages, potentially causing unexpected behavior. The changes include an explicit reset of the isolation level in the connection reset logic and a new test to verify this behavior.

**Bug fix for connection pooling session state:**

* Explicitly reset the transaction isolation level to `READ COMMITTED` in the `Connection::reset()` method to prevent isolation level settings from leaking between pooled connection usages. This addresses the limitation of `SQL_ATTR_RESET_CONNECTION`, which does not reset the isolation level. (`mssql_python/pybind/connection/connection.cpp`)

**Testing improvements:**

* Added a new test `test_connection_pooling_isolation_level_reset` in `tests/test_009_pooling.py` to verify that the isolation level is reset to the default when a connection is reused from the pool, preventing session state leakage.
* Imported the `mssql_python` module in `tests/test_009_pooling.py` to support new test functionality.

### CODE REVIEW

**Review by gargsaumya — APPROVED** [review]

File: mssql_python/pybind/connection/connection.cpp [hunk 1]
```diff
@@ -308,6 +308,19 @@ bool Connection::reset() {
         disconnect();
         return false;
     }
+    
+    // SQL_ATTR_RESET_CONNECTION does NOT reset the transaction isolation level.
+    // Explicitly reset it to the default (SQL_TXN_READ_COMMITTED) to prevent
+    // isolation level settings from leaking between pooled connection usages.
+    LOG("Resetting transaction isolation level to READ COMMITTED");
+    ret = SQLSetConnectAttr_ptr(_dbcHandle->get(), SQL_ATTR_TXN_ISOLATION,
```
Comment (saurabh500) [inline mssql_python/pybind/connection/connection.cpp:316]: What happens for cases where the database option has been changed with alter DB
```
ALTER DATABASE CURRENT
SET READ_COMMITTED_SNAPSHOT ON; 

```

In these cases, I think the snapshot iso level by default changes. If Python sets the ISO to READ COMMITTED, then that likely violates the modified default? (can you verify?)

File: mssql_python/pybind/connection/connection.cpp [hunk 2]
```diff
@@ -308,6 +308,19 @@ bool Connection::reset() {
...
+                                (SQLPOINTER)SQL_TXN_READ_COMMITTED, SQL_IS_INTEGER);
```
Comment (saurabh500) [inline mssql_python/pybind/connection/connection.cpp:317]: When does this set attr on connection get honored? Will this cause an additional roundtrip to SQL server during connection close/open ?

File: mssql_python/pybind/connection/connection.cpp (hunk 1 above)
Comment (sumitmsft) [inline mssql_python/pybind/connection/connection.cpp:316]: What if two threads access the connection pool simultaneously? One could grab this connection while reset() is still executing here, causing race conditions and potential crashes.  

Would it make sense if the entire reset() method should be protected by the connection pool's mutex to prevent concurrent access during reset operations?

Comment (sumitmsft) [reply mssql_python/pybind/connection/connection.cpp:316]: Agreed! Setting isolation level hard coded to Read Committed is not correct. Isolation Level is always a DB property and can be changed by Alter Database... command by a DBA. But at a session level, isolation level can be changed and supersede DB level property - only for that session. Even though the session is finished, DB property will remain as is and should be shown as the default isolation level.

Comment (sumitmsft) [reply mssql_python/pybind/connection/connection.cpp:316]: Having said that, this is the standard default for 99% of SQL Server installations. I have rarely seen this setting being changed frequently. 
DBAs rarely change the server-wide default isolation level - It's an unusual configuration

Comment (sumitmsft) [reply mssql_python/pybind/connection/connection.cpp:316]: I took a dig into this once more. Apparently, there are two things which we need to remember:
1. A DBA can change the isolation level _ONLY_ from READ_COMMITTED to READ_COMMITTED_SNAPSHOT
2. The default isolation level on SQL Server database will always be shown\logged in system metadata as READ_COMMITTED (even though the isolation level is set to READ_COMMITTED_SNAPSHOT). There is another column in system tables (is_snapshot_isolation_enabled) which will be toggled from 0 to 1 if READ_COMMITTED_SNAPSHOT is enabled. 

So, essentially, the isolation level of the DB will also remain READ_COMMITTED. So, even if we hardcode it this will not create any problem. I am ok to go ahead with current changes @subrata-ms

Comment (subrata-ms) [reply mssql_python/pybind/connection/connection.cpp:316]: Thanks @sumitmsft for clarifying this scenario in detail.

File: mssql_python/pybind/connection/connection.cpp (hunk 2 above)
Comment (subrata-ms) [reply mssql_python/pybind/connection/connection.cpp:317]: This issue is only reproducible when we enable the connection pooling in ddbc. The Connection::reset() method will get called when we release the connection to the connection pool. Hence, we need to explicitly reset the isolation level to default ( Read_Committed) otherwise it will hold the previous reset isolation level. Yes, this involves a additional roundtrip to SQL Server.

File: mssql_python/pybind/connection/connection.cpp (hunk 1 above)
Comment (subrata-ms) [reply mssql_python/pybind/connection/connection.cpp:316]: Until Connection::reset() completes it's execution, it will not return to the pool. Hence no other python/query thread get the same connection handle/object. Hence, it is already thread safe through object level. Alternately, the container that holds all the healthy connection is already protected by mutex lock under ConnectionPool::acquire function. Hence current design looks proper to me in this aspect.

Comment (saurabh500) [reply mssql_python/pybind/connection/connection.cpp:316]: @sumitmsft  and @subrata-ms the rare behavior is a non-zero chance. If the feature exists to change the default ISO of a DB, then someone is clearly using it. 

In this case our guidance to folks using non READ_COMMITTED ISO should be to open a connection and explicitly set Isolation level to READ_COMMITTED_SNAPSHOT. Otherwise, a very valid bug saying that the connection pool doesn't honor defaults where the default for DB is READ_COMMITTED_SNAPSHOT can be easily filed. 

We can even do better, have a connection string keyword to specify the isolation level for the connection for customers in this rare scenario. By default we override to READ_COMMITTED and provide a path forward to customers on READ_COMMITTED_SNAPSHOT default databases with connection string override. 

This can be a logical next step to cover the edge cases with this change being brought in. Not a requirement for this PR

File: tests/test_009_pooling.py [hunk 3]
```diff
@@ -229,6 +309,9 @@ def test_pool_idle_timeout_removes_connections(conn_str):
 # =============================================================================
 
 
+@pytest.mark.skip(
```
Comment (saurabh500) [inline tests/test_009_pooling.py:312]: Why skip the test and not delete it?
This seems like a bug which needs to eventually be fixed? 
If yes, it would be good to add a bug number on GH in the skip comment.

**Review by saurabh500 — APPROVED** [review]
Provided some feedback. For its current scope the PR looks good.

File: tests/test_009_pooling.py (hunk 3 above)
Comment (subrata-ms) [reply tests/test_009_pooling.py:312]: There are couple of test that is consistently failing. Had a look and seems it is not proper. I will take this up as part of below task @saurabh500 . 
https://sqlclientdrivers.visualstudio.com/mssql-python/_workitems/edit/40818
